File: frontend/src/core/transaction.js
import { Transaction } from './models.js';
import { transactionValidators } from './validators/index.js';

export function processTransaction(store, transactionData) {
  let violations = [];
  for (const validator of transactionValidators) {
    const result = validator(store, transactionData);
    violations = [...violations, ...result];
  }

  let accountDict = {};
  if (violations.length === 0) {
    const tx = new Transaction(
      transactionData.merchant,
      transactionData.amount,
      transactionData.time
    );
    store.addTransaction(tx);

    const account = store.getAccount();
    if (account) {
      account.availableLimit -= transactionData.amount;
    }
  }

  const account = store.getAccount();
  if (account) {
    accountDict = {
      'active-card': account.activeCard,
      'available-limit': account.availableLimit,
    };
  }

  return [accountDict, violations];
}

File: backend/src/core/transaction.py
from datetime import datetime
from typing import cast
from ..core.models import Transaction
from ..core.state import Store
from ..validators import transaction_validators


def process_transaction(store: Store, transaction_data: dict) -> tuple[dict, list[str]]:
    violations = []
    for validator in transaction_validators:
        result = validator(store, transaction_data)
        violations.extend(result)

    account_dict = {}
    if not violations:
        tx_time = cast(datetime, transaction_data.get("time"))
        transaction = Transaction(
            merchant=transaction_data.get("merchant", ""),
            amount=transaction_data.get("amount", 0),
            time=tx_time
        )
        store.add_transaction(transaction)
        
        account = store.get_account()
        if account:
            account.available_limit -= transaction_data.get("amount", 0)

    account = store.get_account()
    if account:
        account_dict = {
            "active-card": account.active_card,
            "available-limit": account.available_limit
        }

    return account_dict, violations